docs/src/proposals/comprehensive-compute-fees-review.md

```diff
@@ -0,0 +1,164 @@
+---
+title: Comprehensive Compute Fees
+---
+
+## Motivation
+
+The current fee structure lacks a comprehensive account of the work required by
+a validator to process a transaction.  The fee structure is only based on the
+number of signatures in a transaction but is meant to account for the work that
+the validator must perform to validate each transaction.  The validator performs
+a lot more user-defined work than just signature verification.  Processing a
+transaction typically includes signature verifications, account locking, account
+loading, and instruction processing.
+
+## Proposed Solution
+
+The following solution does not specify what native token costs are to be
+associated with the new fee structure.  Instead, it sets the criteria and
+provides the knobs that a cost model can use to determine those costs.
+
+### Fee
+
+The goal of the fees are to prevent the abuse of on-chain compute resources when
+processing  a transaction.  Each of the fee categories below will be represented
+as a compute unit cost that, when added together, encompasses the entire cost of
+processing the transaction.  By calculating the total cost of the transaction,
+the runtime can charge a more representative fee and make better transaction
+scheduling decisions.
+
+A fee will be calculated based on:
+
+1. Number of signatures
+   - Fixed rate per signature
+2. Number of write locks
+   - Fixed rate per writable account
+3. Data byte cost
+   - Fixed rate per byte of the sum of the length all a transactions instruction
+     datas
+4. Account sizes
+   - Account sizes can't be known up-front but can account for a considerable
+     amount of the load the transaction incurs on the network.  The payer will
+     be charged for a maximum account size (10m) upfront and refunded the
+     difference after the actual account sizes are known.
+5. Compute budget
+   - Each transaction will be given a default transaction-wide compute budget of
+     200k units with the option of requesting a larger budget via a compute
```

Comment:
I personally think that transaction-wide budget is the right way to go, but every protocol I worked with combines multiple instructions that add up to over 200K, so such transition should be transparent, i.e. transactions would become more costly one day

Reply:
One of the challenges with making the transition transparent is that we want to have the compute budget information upfront (default or requested larger side) so we can give the cost model better information about how to schedule transaction batches.  When I did an analysis a few months ago there were mainly two offenders of the tx-wide cap, mango and raydium.  What projects are you aware of that are larger?

Reply:
yes, most AMMs fall into this category: Raydium, Saber, Mango (but not Orca, for example, which splits transactions into multiple steps), and protocols that build on top of them, like Tulip or mine :). You don't see it often because it only happens in specific scenarios like the user doesn't have token accounts created or per-user PDAs. I imagine it would be hard to communicate the proposed change to all of the devs, and issues will continue to pop up here and there for a long time. But I don't have enough data, maybe I'm wrong.